File: quant/data/pipeline.py
"""Market data ingestion pipeline orchestrator.

Coordinates fetch -> validate -> store with incremental/full-refresh support.
Designed to run as a daily scheduled job (or on-demand catch-up).

Usage
-----
pipeline = IngestionPipeline(store=MarketDataStore(db_path), source=YahooFinanceSource())
result = pipeline.run(symbols=UNIVERSE, mode="incremental")
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta
from typing import Literal, Sequence

import pandas as pd
from loguru import logger

from quant.data.ingest.base import DataSource, OHLCVRecord
from quant.data.storage.duckdb import MarketDataStore
from quant.data.validation import ValidationResult, validate

# US equity market calendar approximation: Mon–Fri, skipping no holidays
# A proper implementation would use pandas_market_calendars, but we keep
# dependencies minimal here and let gaps indicate non-trading days.
_LOOKBACK_DAYS_INCREMENTAL = 5  # re-fetch last N calendar days for corrections


@dataclass
class PipelineResult:
    symbols_processed: int
    records_fetched: int
    records_valid: int
    records_invalid: int
    records_stored: int
    gaps_detected: dict[str, list[date]]

    def summary(self) -> str:
        gap_count = sum(len(v) for v in self.gaps_detected.values())
        return (
            f"symbols={self.symbols_processed} "
            f"fetched={self.records_fetched} "
            f"valid={self.records_valid} "
            f"invalid={self.records_invalid} "
            f"stored={self.records_stored} "
            f"gaps={gap_count}"
        )


class IngestionPipeline:
    """Fetch, validate, and store OHLCV data.

    Args:
        store: MarketDataStore instance (owns lifecycle).
        source: DataSource implementation to pull from.
        reject_invalid: If True, invalid records are not stored. Default True.
    """

    def __init__(
        self,
        store: MarketDataStore,
        source: DataSource,
        reject_invalid: bool = True,
    ) -> None:
        self._store = store
        self._source = source
        self._reject_invalid = reject_invalid

    def run(
        self,
        symbols: Sequence[str],
        mode: Literal["incremental", "full"] = "incremental",
        start: date | None = None,
        end: date | None = None,
    ) -> PipelineResult:
        """Run the ingestion pipeline.

        Args:
            symbols: Ticker symbols to ingest.
            mode:
                "incremental" — fetch from each symbol's latest stored date
                    (or last 5 calendar days if no history). Efficient for daily runs.
                "full" — fetch entire range defined by start/end.
            start: Used with mode="full". Defaults to 2020-01-01.
            end: Defaults to today.

        Returns:
            PipelineResult with counters and any detected gaps.
        """
        today = date.today()
        end_date = end or today

        symbols = [s.upper() for s in symbols]
        logger.info(
            "Pipeline run: mode={} symbols={} end={}",
            mode,
            len(symbols),
            end_date,
        )

        if mode == "incremental":
            records = self._run_incremental(symbols, end_date)
        else:
            start_date = start or date(2020, 1, 1)
            records = self._run_full(symbols, start_date, end_date)

        # Validate
        vr: ValidationResult = validate(records)
        records_to_store = vr.valid if self._reject_invalid else records

        # Store
        stored = self._store.upsert(records_to_store)

        # Gap detection (per-symbol, last 30 days)
        gaps = self._detect_gaps(symbols, end_date)

        result = PipelineResult(
            symbols_processed=len(symbols),
            records_fetched=len(records),
            records_valid=len(vr.valid),
            records_invalid=len(records) - len(vr.valid),
            records_stored=stored,
            gaps_detected=gaps,
        )
        logger.info("Pipeline complete: {}", result.summary())
        return result

    def _run_incremental(
        self,
        symbols: list[str],
        end_date: date,
    ) -> list[OHLCVRecord]:
        """Fetch from each symbol's latest known date."""
        # Find the earliest "start" we need across all symbols
        fallback_start = end_date - timedelta(days=_LOOKBACK_DAYS_INCREMENTAL)
        min_start = end_date  # will be updated below

        symbol_starts: dict[str, date] = {}
        for sym in symbols:
            latest = self._store.latest_date(sym)
            if latest is None:
                symbol_starts[sym] = fallback_start
            else:
                # Re-fetch from latest - 1 to catch late-arriving corrections
                symbol_starts[sym] = latest - timedelta(days=1)
            if symbol_starts[sym] < min_start:
                min_start = symbol_starts[sym]

        # Batch fetch all with the widest window, then filter per-symbol
        all_records = self._source.fetch(symbols, min_start, end_date)

        # Filter to only include dates >= each symbol's start
        filtered = [
            r for r in all_records
            if r.date >= symbol_starts.get(r.symbol, fallback_start)
        ]
        logger.debug(
            "Incremental: fetched={} filtered={}", len(all_records), len(filtered)
        )
        return filtered

    def _run_full(
        self,
        symbols: list[str],
        start_date: date,
        end_date: date,
    ) -> list[OHLCVRecord]:
        return self._source.fetch(symbols, start_date, end_date)

    def _detect_gaps(
        self,
        symbols: list[str],
        end_date: date,
        lookback_days: int = 30,
    ) -> dict[str, list[date]]:
        """Check for missing business days in the last *lookback_days* calendar days."""
        start = end_date - timedelta(days=lookback_days)
        expected = _business_days(start, end_date)

        gaps: dict[str, list[date]] = {}
        for sym in symbols:
            missing = self._store.coverage_gaps(sym, start, end_date, expected)
            if missing:
                gaps[sym] = missing
                logger.warning(
                    "{}: {} gap(s) detected in last {} days: {}",
                    sym,
                    len(missing),
                    lookback_days,
                    missing[:5],
                )
        return gaps


def _business_days(start: date, end: date) -> list[date]:
    """Return Mon–Fri dates in [start, end]. Does not account for holidays."""
    days: list[date] = []
    current = start
    while current <= end:
        if current.weekday() < 5:  # 0=Mon, 4=Fri
            days.append(current)
        current += timedelta(days=1)
    return days

File: quant-rs/quant-data/src/pipeline.rs
use chrono::{Datelike, NaiveDate, Weekday};

use crate::error::DataError;
use crate::ingest::DataSource;
use crate::models::OhlcvRecord;
use crate::store::MarketDataStore;

/// Summary of a completed ingestion run.
#[derive(Debug, Clone)]
pub struct PipelineResult {
    pub symbols_processed: usize,
    pub records_fetched: usize,
    pub records_stored: usize,
    pub gaps_detected: Vec<(String, Vec<NaiveDate>)>,
}

impl PipelineResult {
    pub fn summary(&self) -> String {
        let gap_count: usize = self.gaps_detected.iter().map(|(_, v)| v.len()).sum();
        format!(
            "symbols={} fetched={} stored={} gaps={}",
            self.symbols_processed,
            self.records_fetched,
            self.records_stored,
            gap_count
        )
    }
}

/// Ingestion mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum IngestMode {
    /// Re-fetch from each symbol's latest stored date (efficient for daily crons).
    Incremental,
    /// Fetch the full date range [start, end].
    Full,
}

const LOOKBACK_DAYS_INCREMENTAL: i64 = 5;

/// Orchestrates fetch → store → gap-detection for OHLCV data.
pub struct IngestionPipeline<S: DataSource> {
    store: MarketDataStore,
    source: S,
}

impl<S: DataSource> IngestionPipeline<S> {
    pub fn new(store: MarketDataStore, source: S) -> Self {
        Self { store, source }
    }

    /// Run the pipeline.
    ///
    /// * `mode = Incremental` — fetch from each symbol's latest date.
    /// * `mode = Full` — fetch `[start, end]`.
    pub fn run(
        &self,
        symbols: &[String],
        mode: IngestMode,
        start: Option<NaiveDate>,
        end: Option<NaiveDate>,
    ) -> Result<PipelineResult, DataError> {
        let today = chrono::Local::now().date_naive();
        let end_date = end.unwrap_or(today);
        let symbols_upper: Vec<String> = symbols.iter().map(|s| s.to_uppercase()).collect();

        let records = if mode == IngestMode::Incremental {
            self.run_incremental(&symbols_upper, end_date)?
        } else {
            let start_date =
                start.unwrap_or_else(|| NaiveDate::from_ymd_opt(2020, 1, 1).unwrap());
            self.source.fetch(&symbols_upper, start_date, end_date)?
        };

        let fetched = records.len();
        let stored = self.store.upsert(&records)?;

        let gaps = self.detect_gaps(&symbols_upper, end_date, 30)?;

        Ok(PipelineResult {
            symbols_processed: symbols_upper.len(),
            records_fetched: fetched,
            records_stored: stored,
            gaps_detected: gaps,
        })
    }

    fn run_incremental(
        &self,
        symbols: &[String],
        end_date: NaiveDate,
    ) -> Result<Vec<OhlcvRecord>, DataError> {
        let fallback = end_date - chrono::Duration::days(LOOKBACK_DAYS_INCREMENTAL);
        let mut min_start = end_date;
        let mut symbol_starts: std::collections::HashMap<&str, NaiveDate> =
            std::collections::HashMap::new();

        for sym in symbols {
            let sym_start = match self.store.latest_date(sym)? {
                Some(latest) => latest - chrono::Duration::days(1),
                None => fallback,
            };
            if sym_start < min_start {
                min_start = sym_start;
            }
            symbol_starts.insert(sym.as_str(), sym_start);
        }

        let all = self.source.fetch(symbols, min_start, end_date)?;
        let filtered: Vec<OhlcvRecord> = all
            .into_iter()
            .filter(|r| {
                r.date
                    >= *symbol_starts
                        .get(r.symbol.as_str())
                        .unwrap_or(&fallback)
            })
            .collect();
        Ok(filtered)
    }

    fn detect_gaps(
        &self,
        symbols: &[String],
        end_date: NaiveDate,
        lookback_days: i64,
    ) -> Result<Vec<(String, Vec<NaiveDate>)>, DataError> {
        let start = end_date - chrono::Duration::days(lookback_days);
        let expected = business_days(start, end_date);
        let mut gaps = Vec::new();

        for sym in symbols {
            let missing = self.store.coverage_gaps(sym, start, end_date, &expected)?;
            if !missing.is_empty() {
                gaps.push((sym.clone(), missing));
            }
        }

        Ok(gaps)
    }
}

/// Return Mon–Fri dates in the inclusive range [start, end].
/// Does not account for public holidays.
pub fn business_days(start: NaiveDate, end: NaiveDate) -> Vec<NaiveDate> {
    let mut days = Vec::new();
    let mut current = start;
    while current <= end {
        match current.weekday() {
            Weekday::Sat | Weekday::Sun => {}
            _ => days.push(current),
        }
        current = current.succ_opt().unwrap_or(current);
        if current == days.last().copied().unwrap_or(current) {
            break;
        }
    }
    days
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ingest::DataSource;
    use crate::models::OhlcvRecord;

    struct MockSource(Vec<OhlcvRecord>);

    impl DataSource for MockSource {
        fn name(&self) -> &str {
            "mock"
        }
        fn fetch(
            &self,
            _symbols: &[String],
            _start: NaiveDate,
            _end: NaiveDate,
        ) -> Result<Vec<OhlcvRecord>, DataError> {
            Ok(self.0.clone())
        }
    }

    fn nd(y: i32, m: u32, d: u32) -> NaiveDate {
        NaiveDate::from_ymd_opt(y, m, d).unwrap()
    }

    fn rec(sym: &str, date: NaiveDate) -> OhlcvRecord {
        OhlcvRecord {
            symbol: sym.to_string(),
            date,
            open: 100.0,
            high: 105.0,
            low: 99.0,
            close: 103.0,
            volume: 1_000_000.0,
            adj_close: 103.0,
        }
    }

    #[test]
    fn full_mode_stores_all_records() {
        let store = MarketDataStore::open(":memory:").unwrap();
        let records = vec![rec("AAPL", nd(2024, 1, 2)), rec("AAPL", nd(2024, 1, 3))];
        let source = MockSource(records);
        let pipeline = IngestionPipeline::new(store, source);

        let symbols = vec!["AAPL".to_string()];
        let result = pipeline
            .run(&symbols, IngestMode::Full, None, Some(nd(2024, 1, 3)))
            .unwrap();

        assert_eq!(result.records_fetched, 2);
        assert_eq!(result.records_stored, 2);
    }

    #[test]
    fn business_days_excludes_weekends() {
        let days = business_days(nd(2024, 1, 1), nd(2024, 1, 7));
        // 2024-01-01 is Monday; 2024-01-06 Sat, 2024-01-07 Sun
        assert!(days.iter().all(|d| {
            !matches!(d.weekday(), Weekday::Sat | Weekday::Sun)
        }));
        assert_eq!(days.len(), 5); // Mon–Fri
    }
}
